File: 0311-sparse-matrix-multiplication/0311-sparse-matrix-multiplication.cpp
class Solution {
private:
    unordered_map<int, vector<int>> flatten(vector<vector<int>>& mat){
        unordered_map<int, vector<int>> m;
        for(int i=0;i<mat.size();i++){
            for(int j=0;j<mat[0].size();j++){
                if(mat[i][j]!=0){
                    m[i].push_back(j);
                }
            }
        }
        return m;
    }
public:
    vector<vector<int>> multiply(vector<vector<int>>& mat1, vector<vector<int>>& mat2) {
        
        auto a = flatten(mat1);
        auto b = flatten(mat2);
        int m = mat1.size(), n = mat1[0].size();
        int k = mat2[0].size();
        vector<vector<int>> res(m, vector<int>(k, 0));
        
        for(auto it: a){
            int row1 = it.first;
            for(auto col1:it.second){
                for(auto it2:b){
                    int row2 = it2.first;
                    for(auto col2:it2.second){
                        if(col1 == row2){
                            res[row1][col2] += mat1[row1][col1] * mat2[row2][col2];
                        }
                    }
                }
            }
        }
        
        return res;
    }
};

File: 0311-sparse-matrix-multiplication/0311-sparse-matrix-multiplication.ts
function flatten(mat: number[][]) : Map<number, number[]> {
    let flatMap = new Map<number, number[]>();
    for(let row=0;row<mat.length;row++){
        flatMap.set(row, new Array<number>());
        for(let col=0;col<mat[0].length;col++){
            if(mat[row][col]!=0) flatMap.get(row).push(col);
        }
    }

    return flatMap;
}

function multiply(mat1: number[][], mat2: number[][]): number[][] {
    let fm1 = flatten(mat1);
    let fm2 = flatten(mat2);
    let m1 = mat1.length, n2 = mat2[0].length;
    
    let res = Array.from(Array(m1), ()=> new Array(n2).fill(0));
    
    for(let [row1,cols1] of fm1.entries()){
        for(let col1 of cols1){
            for(let [row2, cols2] of fm2.entries()){
                for(let col2 of cols2){
                    if(col1 == row2){
                        res[row1][col2] += (mat1[row1][col1] * mat2[row2][col2]);
                    }
                }
            }
        }
    }
    
    return res;
    
};